### PR TITLE
Replica configuration code dedup and refactoring.

### PR DESCRIPTION
**Pull request by toly-kournik**
https://jira.eng.vmware.com/browse/VB-2168

Deduplicate configuration fields which exist in ReplicaConfig from ReplicaImp.
Refactor ReplicaConfigSingleton for optimization and proper initialization.

### CODE REVIEW

File: bftengine/src/bftengine/ReplicaImp.hpp [hunk 1]
```diff
@@ -62,34 +60,31 @@ class StateTransferMsg;
...
 
   // pointers to message handlers
   const std::unordered_map<uint16_t, PtrToMetaMsgHandler> metaMsgHandlers;
 
   // communication
-  class MsgReceiver;  // forward declaration
-  IncomingMsgsStorage incomingMsgsStorage;
-  MsgReceiver* msgReceiver;
+  class MsgReceiver;                                // forward declaration
+  IncomingMsgsStorage incomingMsgsStorage = 20000;  // TODO(GG): use configuration
```
Comment (andrewjstone) [inline bftengine/src/bftengine/ReplicaImp.hpp]: This is a bit strange. You are assigning an integer to a class instance. I assume it's using that in the constructor somehow, but it's not apparent from that line. Maybe change to `incomingMsgsStorage(MAX_INCOMING_MSGS)`?

File: bftengine/tests/testSerialization/TestSerialization.cpp [hunk 2]
```diff
@@ -148,7 +148,7 @@ const uint16_t msgsNum = 2 * fVal + 2 * cVal + 1;
 
 typedef pair<uint16_t, string> IdToKeyPair;
 
-ReplicaConfig config;
+ReplicaConfig *config;
```
Comment (andrewjstone) [inline bftengine/tests/testSerialization/TestSerialization.cpp]: Why dynamically allocate? I also don't see a corresponding delete. Better to use some RAII, so we don't have leaks, no?

File: bftengine/tests/testSerialization/TestSerialization.cpp [hunk 3]
```diff
@@ -425,45 +425,49 @@ void testSetSimpleParams(bool toSet) {
...
+  config->thresholdSignerForExecution = nullptr;
+  config->thresholdVerifierForExecution = new IThresholdVerifierDummy;
+  config->thresholdSignerForSlowPathCommit = new IThresholdSignerDummy;
+  config->thresholdVerifierForSlowPathCommit = new IThresholdVerifierDummy;
+  config->thresholdSignerForCommit = new IThresholdSignerDummy;
+  config->thresholdVerifierForCommit = new IThresholdVerifierDummy;
+  config->thresholdSignerForOptimisticCommit = new IThresholdSignerDummy;
+  config->thresholdVerifierForOptimisticCommit = new IThresholdVerifierDummy;
+  config->singletonFromThis();
+  config->singletonFromThis();
```
Comment (andrewjstone) [inline bftengine/tests/testSerialization/TestSerialization.cpp]: Why do this twice?

Comment (toly-kournik) [reply bftengine/tests/testSerialization/TestSerialization.cpp]: just to check that it works as expected (only once)

File: bftengine/src/bftengine/ReplicaImp.hpp (hunk 1 above)
Comment (toly-kournik) [reply bftengine/src/bftengine/ReplicaImp.hpp]: yes, but this syntax doesn't work

File: bftengine/tests/testSerialization/TestSerialization.cpp (hunk 2 above)
Comment (toly-kournik) [reply bftengine/tests/testSerialization/TestSerialization.cpp]: because ReplicaConfigSingleton holds a pointer to ReplicaConfig

Comment (toly-kournik) [reply bftengine/tests/testSerialization/TestSerialization.cpp]: I have changed an approach a little. Now ReplicaConfigSingleton doesn't hold a pointer to the original ReplicaConfig, but creates a copy, decoupling the life cycles of two objects

**Pull request closed without merging**